### DISCUSSION TITLE
ADCPARAM Problem Thermostat Relay Board (XY-WFT1)

### DISCUSSION THREAD

**Discussion by Prinsessen** [arendst/Tasmota#16656] [Q&A]
Have you looked for this feature in other issues and in the wiki?

Yes

Is your feature request related to a problem? Please describe.

I have installed Tasmota on a  Sinilink Thermostat Relay Board (XY-WFT1), and enabled ANALOG input in source code - that works perfect.

I started out with the default ADCPARAM 2, 32000, 10000, 3350 and the temperature shown in UI is almost accurate.
The problem is that when the temperature supposed to rise it is doing the opposite.

I have measured the resistance on the NTC probe at 25 degree celsius very close to 10K (9890 Ohm)

Then I have calculated the Beta Coefficient using steinhart–hart equation and getting 3508, close to the default.

I have tried with -3508 for the beta coefficient and the temperature is rising as it should, but is very unaccurate at different temperature.

Describe the solution you'd like

I would like to achieve a measurement in the range from 10-80 degree celcius
Hope someone is able to point me in the right direction.

Describe alternatives you've considered

None

**Comment by barbudor** [comment]
What is the schematic for the connection of the NTC ?
The code is made for the Shelly 2.5 NTC Thermistor
```
// 3V3 --- ANALOG_NTC_BRIDGE_RESISTANCE ---v--- NTC --- Gnd
//                                         |
//                                        ADC0
```
If the cabling is different such as having the NTC towards 3V3 and fixed resitor towards GND, this won't work

**Reply by Prinsessen** [reply]
I'm positive it's a pull down resistor, the fixed resistor towords GND.


Wouldn't it be possible to change the code to:

Vout = Vin * R/(Rth+R)

**Reply by barbudor** [reply]
No as it would break the compatibility with Shelly
The only solution is to add an additional ADC mode

Are you using ESP8266 or ESP32 ? official build ?

**Reply by Prinsessen** [reply]
I'm using the ESP8266 and the official build.

**Reply by barbudor** [reply]
I'm looking to use the 4th parameter of AdcParam to specify the alternate cabling

**Reply by barbudor** [reply]
Can you please try with the atteched binary ?
You may need to go 1st to tasmota-minimal.bin.gz before
Once upgraded, enter the same command you were uisng but add `, 1`at the end : `ADCPARAM 2, 32000, 10000, 3350, 1`
This should enable the alternate calculation

If it doesn't work immedtely, I will add more logs but that will be tomorrow. I'm done for today 😄 

tasmota-adctemp_alternate_1.zip (EDIT: link removed as not correct code)

For reference : https://github.com/barbudor/Tasmota/tree/analod_temp_alternate
The major change being the 3 last lines of this block
https://github.com/barbudor/Tasmota/blob/d6cdec003fc2570995057b3d33bd09a35ec0dda5/tasmota/tasmota_xsns_sensor/xsns_02_analog.ino#L498-L506
```C++
      if (Adc[idx].param4) { // Alternate mode
        adc = ANALOG_RANGE - adc;
      }
```

**Reply by Prinsessen** [reply]
Yes it seems to work 😊😊

Just have to find the voltage bridge resistor value - seems alot greater than the default 32000.

Thanks for the quick solution 🤘😊

**Reply by barbudor** [reply]
Good to know
I will PR that tonight

**Reply by handymanny** [reply]
Hey @Prinsessen were you able to figure out the voltage bridge resistor value? I have been fighting with this module for weeks and cannot figure out how to get the temperature to calculate correctly. For me the temperature stays the same and I have actually get a read out from the NTC Thermistor. Could it be that I have done something wrong with the configuration? Or I need to enable something in order to get a value from the Thermistor.

**Comment by Prinsessen** [comment]
Hey @handymanny 

Take a look at this

https://github.com/arendst/Tasmota/discussions/16634#discussioncomment-3718626

**Comment by Prinsessen** [comment]
Or you can flash this, it is working on my board

[tasmota-thermostat.bin.zip](https://github.com/arendst/Tasmota/files/9648868/tasmota-thermostat.bin.zip)

**Reply by barbudor** [reply]
PR has been sent : https://github.com/arendst/Tasmota/pull/16667
It will be soon in official development builds

**Reply by Prinsessen** [reply]
I'm trying to figure out the voltage bridge resistor value .

![ESP8266_pinout](https://user-images.githubusercontent.com/31166607/192460763-d849a9a3-dae8-44b7-9716-e04d05c8d4ea.jpeg)

I'm unable to find any circuit schematic for this board, but it measuring between Vin- to the output pin for the NTC Thermistor I have 10K - and zooming onto the board I have a SMD resistor "01C" that is equal to 10K going to pin 6 on the ESP8086.

When I use Param 2,10000,10000,3940,1 the readout is: 54,4 Degree C (Expected temperature 23 Degree C
When I use Param 2,83000,10000,3940,1 the readout is: 23,6 Degree C (Almost accurate temperature)

The beta coefficient on this NTC thermistor (3940) is checked with the manufactor and measured using Steinhart method, a the NTC Thermistor is 10K and also measured at 25 Degree C

What can I be missing ?

**Reply by barbudor** [reply]
You may not be doing anything wrong
It's just that the original code was mainly made for the Shelly and this board works differently
My change may not be enough

I would suggest you change the gpio17 to standard ADC mode to get direct readings of the ADC and plots a few values for known temperature in order to confront theory with reality

**Reply by Prinsessen** [reply]
I did two plots at:

24 Degree C = 681
100 Degree C = 1024 (I expect I hit a maximum here)

Is this the Vout we see here in mV on pin 6?

NB: In the document it says this board can measure below 60 Degree, would it be of any help to do a measurement at this temperature to see if I get a readout around 1024 also?

**Reply by Prinsessen** [reply]
I did a few more plots:
75 C = 1024
65 C = 1024
61 C = 1024
60 C = 1023
10 C = 474

As this board supports measurements from -40 to 60, these plots seems ok?

**Reply by barbudor** [reply]
1024 is the maximum of the ESP8266 adc when input reach 1V
I think we would need a few points between -40 and +60
A freezer should be around -18

**Reply by Prinsessen** [reply]
The readout at -18 is: 152
I did a reference measuring and its 18 degree

**Reply by Prinsessen** [reply]
-18 degree sorry

**Reply by barbudor** [reply]
https://docs.google.com/spreadsheets/d/1FbNPmiyjOvf3R3Ef4qYY7cezWjvOzcwsQxUqv1s_4XE/edit?usp=sharing

What is the value of the other resistor ? the one towards GND ?
I read you are using value 32000. Are you sure it is 32Kohms ?

**Reply by barbudor** [reply]
Trying to calculate the value of the NTC in ohms but is the bridge resistor is 32K, the value I get are not consistent with your measured value of ~10K at 24C

**Reply by barbudor** [reply]
~20K would be more logical as NTC@20C is measured at 9890 ohms, and at 24C, ADC voltage is 0.666 = NTC / (NTC + R)
So NTC@24C = 1/2 R

**Reply by Prinsessen** [reply]
I'm positive that the bridge resistor is 10k.
I tried with 20000 and that gives me 60C when it should be 24.

Adcparam 2,20000,10000,3910,1

But 10k gives me a even higher temperature, that's what confused me. 

83000 83K gives me the correct temperature, but in top and button of the scale it's complete wrong.

**Reply by barbudor** [reply]
1st time you say the reistor is 10K 😄 
All exemples above you wrote 32000

So now I have this:
![image](https://user-images.githubusercontent.com/2996042/192626247-926e9ef0-3e60-493c-a779-a6022e4763fb.png)

Does those NTC values looks reasonable ?

If yes, I can then try to understand what goes wrong in the calculations

**Reply by Prinsessen** [reply]
The NTC is a 10k 25C that is taken from the datasheet and is also measured, disconnected from the board and put into 25C water.

**Reply by Prinsessen** [reply]
No I wrote in a early replay this 

"" 
I'm unable to find any circuit schematic for this board, but it measuring between Vin- to the output pin for the NTC Thermistor I have 10K - and zooming onto the board I have a SMD resistor "01C" that is equal to 10K going to pin 6 on the ESP8086."

But using 10000 was giving me a wrong temperature, a resistor value in adcparam of 83000 was giving me the most accurate temperature but not in the outer temperature range.

**Reply by Prinsessen** [reply]
The math for an NTC thermistor with a pull down resistor I believe is this :

Vout = Vin * R/(Rth+R)

**Reply by barbudor** [reply]
I think I understood the equations and see what is the problem with my solution
I'll be back soon

**Reply by barbudor** [reply]
So I took back my pencil and a paper ...
The problem was that on ESP8266 the ADC full-scale is 1.0 V while the bridge Rt/R is across 3V3
So my trick of doing `adc = ANALOG_RANGE - adc` was wrong (altough it should have worked on ESP32 as the full scale is 3V3)
So the new code is better as I change the calculation of Rt by re-doing the formulas properly:
```C++
#ifdef ESP8266
      if (Adc[idx].param4) { // Alternate mode
        Rt = (double)Adc[idx].param1 * (ANALOG_RANGE * ANALOG_V33 - (double)adc) / (double)adc;
      } else {
        Rt = (double)Adc[idx].param1 * (double)adc / (ANALOG_RANGE * ANALOG_V33 - (double)adc);
      }
#else
      if (Adc[idx].param4) { // Alternate mode
        Rt = (double)Adc[idx].param1 * (ANALOG_RANGE - (double)adc) / (double)adc;
      } else {
        Rt = (double)Adc[idx].param1 * (double)adc / (ANALOG_RANGE - (double)adc);
      }
#endif
```

Can you please try this one ?
[tasmota-adctemp_alternate_2.zip](https://github.com/arendst/Tasmota/files/9659586/tasmota-adctemp_alternate_2.zip)

**Comment by Prinsessen** [comment]
That look much better :) 

I do have to find the correct value for the Pulldown resistor - I think i could use the equation to do so.
It's impossible to track the resistor on the PCB board, and there is no print layout to be found.
With a Pull-down resistor 2,2k it seems ok now, with a little fine tuning of the beta coefficient and the R.

Can you please share the : tasmota/tasmota_xsns_sensor/xsns_02_analog.ino so I don't do any stupid mistake compiling?

**Reply by Prinsessen** [reply]
I did some test measuring with the new code.

I have 3 factors for the equation: Rt, R, Beta .
Two of them I'm 100% sure of (Rt and Beta), but R is still the troubled one.

With 10k resistor I'm now getting -4.9C at 22C.
With  2,2k resistor i'm now getting 25.1C at 22C

But still problem in the top and bottom range

**Comment by barbudor** [comment]
Latest code is there : https://github.com/barbudor/Tasmota/blob/fix-adc_temp-alternate-mode/tasmota/tasmota_xsns_sensor/xsns_02_analog.ino

**Comment by Prinsessen** [comment]
I checked the equation and solved it for RTh:

RTh=(-R*Vout+R*Vin)/Vout

Is that the same you use in your code?

**Reply by barbudor** [reply]
Yes
Given that 
adc = the 0..1023 value read from the ADC, with a full scale of 1.0V on ESP8266
Vout = adc / ANALOG_RANGE
Vin = ANALOG_3.3 = 3.3 : The voltage at the top of the bridge

RTh = (-RVout+RVin)/Vout 
Rth = R . (Vin - Vout)/Vout
Rth = R . (ANALOG_3.3 - adc / ANALOG_RANGE) / (adc / ANALOG_RANGE)
Rth = R . (ANALOG_3.3 * ANALOG_RANGE - adc / ANALOG_RANGE) / adc

Which is the formula here:
Rt = (double)Adc[idx].param1 * (ANALOG_RANGE * ANALOG_V33 - (double)adc) / (double)adc;

**Comment by Prinsessen** [comment]
`Rth = R . (ANALOG_3.3 * ANALOG_RANGE - adc / ANALOG_RANGE) / adc`

Don't we need to Divide with ANALOG_RANGE aswell, or have I misunderstood?

`Rt = (double)Adc[idx].param1 * (ANALOG_RANGE * ANALOG_V33 - (double)adc / ANALOG_RANGE ) / (double)adc;`

**Comment by Prinsessen** [comment]
I was thinking that they might have done something special around the pull-down resistor so my guess is that I would be able to get closer, calculating R:

Given:


```
25C
ADC reading 663
VOut=663/1024=0,6474v
NTC25C=10000

R=(RTh*Vout) / (-Vout+Vin)

R=10000*0,6474 / -0,6474+3,3

R=2440 Ohm
```
With this:

adcparam 2,2440,10000,3940,1

I get a readout with the code at 23,6 C (My reference thermometer says 22,8)

If i disconnect the NTC (ADC Read 0) readout says -43,1 C
If i short the NTC (ADC Read 1024) readout says 38,7 C

I simply can't say if I do something wrong or a error in the code

**Reply by barbudor** [reply]
Unfortunately I have no idea if the code was working well before on Shelly
I think I have a NTC somewhere in my part bin but not sure I will have the coefficient
I will try to setup a test bench tonight and if proven bad, I'll try to look further in the implementation of the S-H formula

Out of curiosity, is there an interrest in using an NTC vs a DS18X20 ?

**Reply by Prinsessen** [reply]
Something is not right here - hope we can work it out :)

My application for the Sinilink is to act as Thermostat for seat heat on my Indian Scout Bobber

![Bobber](https://user-images.githubusercontent.com/31166607/192750316-27923a10-9cb2-4422-a0b7-a75058da249e.jpeg)

The NTC Thermistor is already build in the seat, so a change of probe needs an opening of the seat :)

**Reply by Jason2866** [reply]
Nice! I ordered one too. Not the bike ;-)

**Reply by Jason2866** [reply]
> Unfortunately I have no idea if the code was working well before on Shelly I think I have a NTC somewhere in my part bin but not sure I will have the coefficient I will try to setup a test bench tonight and if proven bad, I'll try to look further in the implementation of the S-H formula

@barbudor I think i read some statements from users in Discord about strange temperature readings from Shelly. I am still not sure if the formula for esp8266 is fully correct (even after the changes you made).

**Reply by Prinsessen** [reply]
If its one of those cheap Ebay, Aliexpress ones, they usually are with a coefficient at 3950

**Reply by barbudor** [reply]
Took me a little time to have something that kind works
I have a 4.7K between A0 and GND. As it is a mini-D1 there is on board 100K also. Measured at 4490 ohms.
The NTC is supposed ot 10K, 3950 (I found the old Ali order)
I put a Mija BLE sensor close to the NTC
Setup `AdcParam 2, 4490, 10000, 3950, 1`
ADC reads around 937..945
Rt computed around 11600
Temperature displayed : 21.7°C, for 19.7 displayed on the Mija

Trying now to understand the formula

**Reply by barbudor** [reply]
I have pushed https://github.com/arendst/Tasmota/pull/16702 which fixes the previous calculations for Sinilink mode. That's just the same as tasmota-adctemp_alternate_2.zip from 2 days ago.

However I still don't know how to improve accuracy.
Is it coming from the ADC ?
Is it coming form the formula used ?

The formula is a simplified Steinhart-Hart formula using only B-coefficient and a single log:

![image](https://user-images.githubusercontent.com/2996042/193362122-66bd9c5e-8ea7-4294-b57d-732d9cbd8dc7.png)

To use the full formula would require also to have the C-coefficient:

![image](https://user-images.githubusercontent.com/2996042/193362523-80a2fc47-d0a8-4e5b-8233-95026b6a6e72.png)

Formula pictures taken from  https://arduinodiy.wordpress.com/2015/11/10/measuring-temperature-with-ntc-the-steinhart-hart-formula/

**Reply by Prinsessen** [reply]
The accuracy in my setup is:

AdcParam 2, 2700, 10000, 3950, 1

At 25 Centigrade, there is a difference between sinilink and reference probe of 2 degree.

At 35 Centigrade this gab is almost 8 degree.

I'll test to se if the error could lie in the ref probe, or it has to be in the Steinheart calculation something is wrong.

**Comment by lalo-uy** [comment]
After looking int the code an the equations I belive the ANALOG_33 has no thing to do there.
The relatiotins Vi/Vcc is the same as adc/ADC_RANGE, we never expressed nothing in volts.
Will try to compile my test fw.
